File: solutions/java/code-carl/ch02_list/T4_LC0206_ReverseList.java
/**
 * Practice template for:
 * - LeetCode 206. Reverse Linked List
 *   https://leetcode.cn/problems/reverse-linked-list/
 */
public class T4_LC0206_ReverseList {
    public ListNode reverseList(ListNode head) {
        ListNode curr = null;
        ListNode post = head;
        while (post != null) {
            ListNode tmp = post.next;
            post.next = curr;
            curr = post;
            post = tmp;
        }
        return curr;
    }
}

File: archive/2026-09-18/ch02_list/t4_lc0206_reverse_list.py
"""LeetCode 206. Reverse Linked List
https://leetcode.cn/problems/reverse-linked-list/
"""

from typing import Optional

from list_node import ListNode


class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        pre, curr = None, head
        while curr:
            tmp = curr.next
            curr.next = pre
            pre = curr
            curr = tmp

        return pre


if __name__ == "__main__":
    from list_utils import build_list, to_list

    solution = Solution()
    assert to_list(solution.reverseList(build_list([1, 2, 3, 4, 5]))) == [5, 4, 3, 2, 1]
    assert solution.reverseList(None) is None
    print("All tests passed.")
